### ISSUE TITLE
batch内负样本采样

### ISSUE THREAD

**Issue by aph-asic** [NLPJCL/RAG-Retrieval#24]
![image](https://github.com/NLPJCL/RAG-Retrieval/assets/54587975/baad5cc8-0080-41fc-add3-b5fb199bf766)

你好，对于batch内随机采样负样本，可能存在某样本的负样本刚好是同batch内一个样本的正样本，如果采样到了，是否可能会对模型效果造成影响了，如果要采用这种随机采样的方法，是否要求数据集的数量够大

**Comment by NLPJCL** [comment]
可能存在某样本的负样本刚好是同batch内一个样本的正样本，如果采样到了，是否可能会对模型效果造成影响了。
1.是的，会存在这种现象，但是如果数据随机打乱，并且doc侧数量较大的话，这种发生的概率较小。

**Comment by aph-asic** [comment]
> 可能存在某样本的负样本刚好是同batch内一个样本的正样本，如果采样到了，是否可能会对模型效果造成影响了。 1.是的，会存在这种现象，但是如果数据随机打乱，并且doc侧数量较大的话，这种发生的概率较小。

        if use_inbatch == True:
        # [batch_size,1,batch_size*neg_nums] <- [batch_size,1,dim],[1,batch_size*neg_nums,dim]
            neg_sim_matrix = query_embeddings.unsqueeze(1) @ neg_doc_embeddings.unsqueeze(0).transpose(-1, -2)
            neg_sim_matrix = neg_sim_matrix.squeeze(1)
            labels = torch.zeros(query_embeddings.shape[0], dtype=torch.long, device=query_embeddings.device)
            pos_neg_score = torch.cat([pos_sim_matrix.unsqueeze(1), neg_sim_matrix], dim=1) / self.temperature

        else:
            neg_sim_matrix = torch.sum(query_embeddings * neg_doc_embeddings, dim=-1)
            labels = torch.zeros(query_embeddings.shape[0], dtype=torch.long, device=query_embeddings.device)
            pos_neg_score = torch.cat([pos_sim_matrix.unsqueeze(1), neg_sim_matrix.unsqueeze(1)], dim=1) / self.temperature

能麻烦您帮我看下这个代码是否正确吗

**Comment by NLPJCL** [comment]
看起来没有问题。
实现的是，embedding模型，输入是query-pos-neg的时候，不使用batch内其他query的负例作为当前query的随机负例。

**Comment by aph-asic** [comment]
> 看起来没有问题。 实现的是，embedding模型，输入是query-pos-neg的时候，不使用batch内其他query的负例作为当前query的随机负例。
谢谢

**Comment by HuXiLiFeng** [comment]
```
if self.use_inbatch == True:
    # [batch_size,1,batch_size*neg_nums] <- [batch_size,1,dim],[1,batch_size*neg_nums,dim]
    neg_sim_matrix = query_embeddings.unsqueeze(1) @ neg_doc_embeddings.unsqueeze(0).transpose(-1, -2)

    # [batch_size,batch_size*neg_nums]
    neg_sim_matrix = neg_sim_matrix.squeeze(1)
else:
    # [batch_size, neg_nums, dim]
    neg_doc_embeddings = torch.reshape(neg_doc_embeddings, (query_embeddings.shape[0], -1, neg_doc_embeddings.shape[-1]))
    # [batch_size, neg_nums]
    neg_sim_matrix = torch.sum(neg_doc_embeddings * query_embeddings.unsqueeze(1), dim=-1)
```

参考下这个呢
